### PR TITLE
Reset before initializer and optimization to remove it when zero state 

### LINKED ISSUE #1896: Initialize instruction should unroll to reset + gates
### What is the expected enhancement?

Make `initialize` a gate element that can be used directly in Aer simulators, and unrolled into reset + gates by the transpiler for devices and noisey simulations. 

#### Current version
The (current) format for initialize as a `QuantumCircuit` method is `circ.initialize(statevec, qr)` to initialise a quantum register in a fixed state. 

### Proposed changes

* Change `intialize` it to accept either single-qubits or a list (or register) of qubits and to be treated as an explicit reset operation followed by gates to prepare the input state.
* On simulators and devices that support reset this would allow `initialize` to be placed anywhere in a circuit acting on a subset of qubits.
* To simplify the transpiler could remove unneeded reset operations.
   * If a qubit has more than 1 reset operation in a row
   * If reset is determined to be the first operation on a given qubit (*Note: but for noisy simulations one might want to leave explicit reset operations at the start of a circuit to allow for reset errors*)

#### Circuit method

The circuit method could be `initialize(state, qubits)`, where `state` is a `list/ndarray` that is a valid statevector, and `qubits` can be either a full quantum register, single-qubit, or list/tuple of qubits that match the dimension of the state vector.

#### Qobj serialization

A suggestion for how this could be directly serialized to qobj (if the backend supports the "initialize" instruction in its basis gates) is

```
{"name": "initialize", "qubits": list[int], "params": complex_vector}
```  

### Example:

Consider a 2-qubit circuit, where I want to initialize both qubits into the `|1>` state, the following should all be equivalent:

```python
qr = QuantumRegister(2)

# Method 1
circ1 = QuantumCircuit(qr)
circ1.initialize([0, 0, 0, 1], qr)

# Method 2
circ2 = QuantumCircuit(qr)
circ2.initialize([0, 0, 0, 1], (qr[0], qr[1]))

# Method 3
circ3 = QuantumCircuit(qr)
circ3.initialize([0, 1], qr[0])
circ3.initialize([0, 1], qr[1])


# Implementation
circ0 = QuantumCircuit(qr)
circ0.reset(qr[0])
circ0.x(qr[0])
circ0.reset(qr[1])
circ0.x(qr[1])
```

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #1896.

This PR adds a reset at the begin of `InitializeGate`. Also extend the `default_pass_manager_simulator` to remove the gate when its at the beginning of the circuit.

### CODE REVIEW

File: qiskit/extensions/quantum_initializer/initializer.py [hunk 1]
```diff
@@ -72,8 +73,8 @@ def _define(self):
 
         q = QuantumRegister(self.num_qubits, 'q')
         initialize_circuit = QuantumCircuit(q, name='init_def')
-        # TODO: make initialize an Instruction, and insert reset
```
Comment (ajavadia) [inline qiskit/extensions/quantum_initializer/initializer.py]: Can you derive Initialize from `Instruction`? Since it is not a Gate anymore (gates assumed to be reversible).
Also it should probably be called just `Initialize` now, rather than `InitializeGate`.

File: qiskit/transpiler/preset_passmanagers/default.py [hunk 2]
```diff
@@ -98,6 +100,9 @@ def default_pass_manager_simulator(basis_gates):
         PassManager: A passmanager that just unrolls, without any optimization.
     """
     pass_manager = PassManager()
+
+    pass_manager.append(Decompose(InitializeGate))
```
Comment (ajavadia) [inline qiskit/transpiler/preset_passmanagers/default.py]: Shouldn't need this, because the `Unroller` pass will automatically decompose the initialize instruction if `initialize` is not in the backend's basis gates.

In fact we should not force decomposition here. Because simulators usually support direct initialization, which is much faster than decomposing.

File: qiskit/transpiler/preset_passmanagers/default.py [hunk 3]
```diff
@@ -98,6 +100,9 @@ def default_pass_manager_simulator(basis_gates):
...
+    pass_manager.append(RemoveResetInZeroState())
```
Comment (ajavadia) [inline qiskit/transpiler/preset_passmanagers/default.py]: This has to be in the Default pass manager too, not just for the simulators.

**Comment by 1ucian0** [comment]
Any comment about the failing test ?
https://travis-ci.com/Qiskit/qiskit-terra/jobs/195909610#L2456

File: qiskit/extensions/quantum_initializer/initializer.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/extensions/quantum_initializer/initializer.py]: If you dont mind, I would prefer to do that refactor in a different PR, called "Initialize is an instruction, not a gate". I prefer small PRs and the refactor is not directly related with #1896.

**Comment by ajavadia** [comment]
Ok the failing test was actually a wrong test! And it was because we were not resetting. So now the behavior is correct, and so the test should be modified like this:
```python
     def test_combiner(self):
         """Combining two circuits containing initialize."""
-        desired_vector = [0, 1]
+        desired_vector_1 = [1.0 / math.sqrt(2), 1.0 / math.sqrt(2)]
+        desired_vector_2 = [1.0 / math.sqrt(2), -1.0 / math.sqrt(2)]
         qr = QuantumRegister(1, "qr")
         cr = ClassicalRegister(1, "cr")
         qc1 = QuantumCircuit(qr, cr)
@@ -271,7 +272,7 @@ class TestInitialize(QiskitTestCase):
         job = execute(qc1 + qc2, BasicAer.get_backend('statevector_simulator'))
         result = job.result()
         quantum_state = result.get_statevector()
-        fidelity = state_fidelity(quantum_state, desired_vector)
+        fidelity = state_fidelity(quantum_state, desired_vector_2)
         self.assertGreater(
             fidelity, self._desired_fidelity,
             "Initializer has low fidelity {0:.2g}.".format(fidelity))
```

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Fixed in cbc7fb5d

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Fixed in 52e02902

**Comment by 1ucian0** [comment]
Okey. Test fixed. Removing [WIP]. Let me know what you think!

File: qiskit/transpiler/preset_passmanagers/default.py [hunk 4]
```diff
@@ -81,6 +82,9 @@ def default_pass_manager(basis_gates, coupling_map, initial_layout,
     else:
         simplification_passes = [CXCancellation()]
 
+    if 'reset' in basis_gates:
```
Comment (ajavadia) [inline qiskit/transpiler/preset_passmanagers/default.py]: can you remove this? this optimization is not related to having reset in the basis.

File: qiskit/transpiler/preset_passmanagers/default.py [hunk 5]
```diff
@@ -98,6 +102,11 @@ def default_pass_manager_simulator(basis_gates):
         PassManager: A passmanager that just unrolls, without any optimization.
     """
     pass_manager = PassManager()
+
     pass_manager.append(Unroller(basis_gates))
 
+    if 'reset' in basis_gates:
```
Comment (ajavadia) [inline qiskit/transpiler/preset_passmanagers/default.py]: same here.

File: qiskit/extensions/quantum_initializer/initializer.py (hunk 1 above)
Comment (ajavadia) [reply qiskit/extensions/quantum_initializer/initializer.py]: It's just two lines I think. But this PR is inserting a reset in initialize, so it is converting it from a gate to an instruction, so it seems right that it becomes an Instruction in this PR.

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Otherwise the pass does nothing, because it searches for the `Reset` gate after unrolling.

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Same reason :)

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 4 above)
Comment (ajavadia) [reply qiskit/transpiler/preset_passmanagers/default.py]: I think there's a confusion here between `basis_gates` and the gates inside a circuit.
The `basis_gates` are what is supported by the backend. The circuit may have reset, and the backend may not support it (in fact this is what usually happens). So we want this pass to run on all *circuits* that have reset in them. 

The pass is correct. If you just remove this if statement, it should work. It will automatically not do anything for circuits that don't have reset (which is obviously correct, there is nothing to remove).

**Comment by ajavadia** [comment]
The file location of the intializer was changed in #2173, so this branch needs to be rebased.

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Sure. I'm talking in the contest of the default pipeline. `RemoveResetInZeroState` is wasting cycles if there is no `Reset` in the circuit. If `RemoveResetInZeroState` runs after unrolling to the basis, then `Reset` needs to be in the basis for `RemoveResetInZeroState` to do something. This is only true in the particular situation of the this particular pipeline.

Comment (ajavadia) [reply qiskit/transpiler/preset_passmanagers/default.py]: No that's not true, just add this test and you'll see what I mean:
```
import math
from qiskit import QuantumCircuit, execute
circuit = QuantumCircuit(2, 2)

desired_vector = [1 / math.sqrt(2), 1 / math.sqrt(2), 0, 0]

circuit.initialize(desired_vector, range(0,2))
circuit.measure(range(0,2), range(0,2))
print(circuit)

from qiskit.test.mock import FakeMelbourne
from qiskit.compiler import transpile
device_backend = FakeMelbourne()
new_circuit = transpile(circuit, device_backend)
print(new_circuit)
```
Here the reset should be removed, but the way you've done it now leaves the reset in.
Reset is just ignored by the unroller if you look at the unroller code.

Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: you are right. I missed the introduction of `basic_insts`. I will remove the conditional. BTW, what this `TODO` means?
```python
            basic_insts = ['measure', 'reset', 'barrier', 'snapshot']
            if node.name in basic_insts:  # TODO: this is legacy behavior
                continue
```

Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Done in 79fa7d71

File: qiskit/transpiler/preset_passmanagers/default.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/transpiler/preset_passmanagers/default.py]: Done in 79fa7d71

File: qiskit/extensions/quantum_initializer/initializer.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/extensions/quantum_initializer/initializer.py]: I tried... no 2 lines, but doable with little changes :) Done in 3ada9dd47b459bb092b485dc7fd751415ed04858
